### ISSUE TITLE
[supposed bug] Compiler  set a .php file in the xml installer

### ISSUE THREAD

**Issue by jcodewalker** [vdm-io/Joomla-Component-Builder#454]
Hello,

JCB version 3.9.21 was setting the file componentname.php twice in the xml installation file, in the site side, something like:
`	<files folder="site">
<filename>controller.php</filename>
<filename>index.html</filename>
<filename>router.php</filename>
<filename>componentname.php</filename>
<filename>componentname.php</filename>
<filename>router.php</filename>
<filename>controller.php</filename>
<folder>assets</folder>
<folder>helpers</folder>
<folder>controllers</folder>
<folder>models</folder>
<folder>views</folder>
<folder>layouts</folder>
</files>
`
I didn't know until now as the installation worked fine since it was in the the same site of JCB, where the component was developed.

Today I wanted to export that package and import it in a brand new installation (both Joomla (3.9.10) and JCB (2.9.35).
But once imported I was not able to install it (after compiled fine). The error was "file doesn't exists blablabla/site/.php".
Checking into the xml installation file I found:
`	<files folder="site">
		<filename>controller.php</filename>
		<filename>index.html</filename>
		<filename>router.php</filename>
		<filename>.php</filename>
		<filename>componentname.php</filename>
		<filename>router.php</filename>
		<filename>controller.php</filename>
		<folder>assets</folder>
		<folder>helpers</folder>
		<folder>controllers</folder>
		<folder>models</folder>
		<folder>views</folder>
		<folder>layouts</folder>
	</files>
`
This is when I found the issue. I wanted to compare the compiled archives and I found the difference: JCB 2.9.21 set twice the component name, while 2.9.35 set a '.php' file and the componentname.php.

I've updated the older installation to JCB 2.9.35 and now it doesn't set twice the componentname.php file but it set a '.php' file as well, like:
`	<files folder="site">
		<filename>controller.php</filename>
		<filename>index.html</filename>
		<filename>router.php</filename>
		<filename>.php</filename>
		<filename>componentname.php</filename>
		<filename>router.php</filename>
		<filename>controller.php</filename>
		<folder>assets</folder>
		<folder>helpers</folder>
		<folder>controllers</folder>
		<folder>models</folder>
		<folder>views</folder>
		<folder>layouts</folder>
	</files>
`
Result, now I'm not able to install the component in the "old" site (where it has been developed) nor in the brand new one.

Any hint?
Thanks in advance!

**Comment by Llewellynvdm** [comment]
I just noticed this last night when I wanted to start with some of those tutorials. So it has been fixed, and should in the next release be resolved. To patch it you must update a few lines in the compiler.

after [line 243 in the b_Structure.php](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/b_Structure.php#L243) file add:
```php
/**
 * The standard root files
 * 
 * @var      array
 */
public $stdRootFiles = array('access.xml', 'config.xml', 'controller.php', 'index.html', 'README.txt');
```
Then in the [constructor after line 335](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/b_Structure.php#L335) add:
```php
$this->stdRootFiles[] = $this->componentData->name_code . '.php';
```
Then replace all code from line [892](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/b_Structure.php#L892) up to [line 926](https://github.com/vdm-io/Joomla-Component-Builder/blob/master/admin/helpers/compiler/b_Structure.php#L926) with the following code:
```php
// check if we should add the dynamic folder moving script to the installer script
$checker = array_values((array) explode('/', $zipFullPath));
// TODO <-- this may not be the best way, will keep an eye on this.
// We basicly only want to check if a folder is added that is not in the stdFolders array
if (isset($checker[0]) && ComponentbuilderHelper::checkString($checker[0])
	&& !in_array($checker[0], $this->stdFolders))
{
	// check if we should add the dynamic folder moving script to the installer script
	if (!$this->setMoveFolders)
	{
		// add the setDynamicF0ld3rs() method to the install scipt.php file
		$this->setMoveFolders = true;
		// set message that this was done (will still add a tutorial link later)
		$this->app->enqueueMessage(JText::_('<hr /><h3>Dynamic folder/s were detected.</h3>'), 'Notice');
		$this->app->enqueueMessage(JText::sprintf('A method (setDynamicF0ld3rs) was added to the install <b>script.php</b> of this package to insure that the folder/s are copied into the correct place when this componet is installed!'), 'Notice');
	}
}
elseif (count($checker) == 2 && ComponentbuilderHelper::checkString($checker[0]))
{
	$add_to_extra = false;
	// set the target
	$eNAME = 'FILES';
	$ename = 'filename';
	// this should not happen and must have been caught by the above if statment
	if ($details->type === 'folder')
	{
		// only folders outside the standard folder are added
		if (!in_array($checker[0], $this->stdFolders))
		{
			$eNAME = 'FOLDERS';
			$ename = 'folder';
			$add_to_extra = true;
		}
	}
	// if this is a file, it can only be added to the admin/site/media folders 
	// all other folders are moved as a whole so their files do not need to be declared
	elseif (in_array($checker[0], $this->stdFolders) && !in_array($checker[1], $this->stdRootFiles))
	{
		$add_to_extra = true;
	}
	// add if valid folder/file
	if ($add_to_extra)
	{
		// set the tab
		$eTab = $this->_t(2);
		if ('admin' === $checker[0])
		{
			$eTab = $this->_t(3);
		}
		// set the xml file
		$this->fileContentStatic[$this->hhh . 'EXSTRA_' . ComponentbuilderHelper::safeString($checker[0], 'U') . '_' . $eNAME . $this->hhh] .= PHP_EOL . $eTab . "<" . $ename . ">" . $checker[1] . "</" . $ename . ">";
	}
}				
```
This should fix the duplication, we are going to move this area latter to follow the new path we are using for the new plugin area. But this will work for now.

**Comment by Llewellynvdm** [comment]
Just another head-up the line numbers are in relation to the master branch at the current date.

O yes and then it must be that you are using the standing branch which I would advice against... this when working on your official components. Always use only the master branch... since the behavior of the only `.php` and many other related missing component names are part of the staging branch errors we also found in the last push. This fix (in the staging branch) replace the whole method called `getGlobalPlaceholders` in the [a_Get.php](https://github.com/vdm-io/Joomla-Component-Builder/blob/staging/admin/helpers/compiler/a_Get.php#L932) file with the following:

```php
	/**
	 * get all System Placeholders
	 *
	 * @return  array The global placeholders
	 * 
	 */
	public function getGlobalPlaceholders()
	{
		// reset bucket
		$bucket = array();
		// Create a new query object.
		$query = $this->db->getQuery(true);
		$query->select($this->db->quoteName(array('a.target','a.value')));
		// from these tables
		$query->from('#__componentbuilder_placeholder AS a');
		// Reset the query using our newly populated query object.
		$this->db->setQuery($query);
		// Load the items
		$this->db->execute();
		if ($this->db->getNumRows())
		{
			$bucket = $this->db->loadAssocList('target', 'value');
			// open all the code
			foreach ($bucket as $key => &$code)
			{
				$code = base64_decode($code);
			}
		}
		// set component place holders
		$bucket[$this->hhh . 'component' . $this->hhh] = $this->componentCodeName;
		$bucket[$this->hhh . 'Component' . $this->hhh] = ComponentbuilderHelper::safeString($this->componentCodeName, 'F');
		$bucket[$this->hhh . 'COMPONENT' . $this->hhh] = ComponentbuilderHelper::safeString($this->componentCodeName, 'U');
		$bucket[$this->bbb . 'component' . $this->ddd] = $bucket[$this->hhh . 'component' . $this->hhh];
		$bucket[$this->bbb . 'Component' . $this->ddd] = $bucket[$this->hhh . 'Component' . $this->hhh];
		$bucket[$this->bbb . 'COMPONENT' . $this->ddd] = $bucket[$this->hhh . 'COMPONENT' . $this->hhh];
		// get the current components overides
		if (($_placeholders = ComponentbuilderHelper::getVar('component_placeholders', $this->componentID, 'joomla_component', 'addplaceholders')) !== false
			&&  ComponentbuilderHelper::checkJson($_placeholders))
		{
			$_placeholders = json_decode($_placeholders, true);
			if (ComponentbuilderHelper::checkArray($_placeholders))
			{
				foreach($_placeholders as $row)
				{
					$bucket[$row['target']] = $row['value'];
				}
			}
		}
		return $bucket;
	}
```

**Comment by Llewellynvdm** [comment]
The staging branch is like nightly builds and are unstable and full of bugs... not ready for live development. But okay it also where all the latest ideas are already active... hope you manage, let me know.

**Comment by Llewellynvdm** [comment]
Okay I just pushed the fix up and this issue should be resolved in the staging branch :+1:

**Comment by jcodewalker** [comment]
Hello Llewellyn,

Thanks for your fast reply.

Yes it's my fault, I downloaded the staging as the switch didn't switched (or I clicked on staging meaning master...) and I didn't check it.

So, after the patch, now I can install it on the site where it was developed and even in the other one, where I've imported the mapped package.
However, the xml install file still has the componentname.php set twice...
Same result in both sites.
It seems doesn't prevent the installation so it's just an info for your fixing.

Thanks a lot!
